fud/fud/json_to_dat.py

```diff
@@ -2,34 +2,108 @@
     val &= (2**bw - 1)
     return '{:x}'.format(val)
 
+def bitstring2(val, bw):
+    #first truncate val by `bw` bits
+    val = val[len(val)-bw:len(val)]
+    numval = int(val,base=2)
+    return '{:x}'.format(numval)
+
 
 def parse_dat(path):
     with path.open('r') as f:
         lines = f.readlines()
         arr = np.array(list(map(lambda v: int(v.strip(), 16), lines)))
         return arr
 
+def parse_dat_fxd(path, wholebit, intbit, fracbit):
+    with path.open('r') as f:
+        lines = f.readlines()
+        arr = np.array(
+            list(
+                map(
+                    lambda v: 
+                    fixed_p_to_decimal(np.binary_repr(int(v.strip(), 16),width = wholebit),
+                    wholebit, intbit,fracbit), lines)))
+        return arr
+
+def decimal_to_fixed_p (num, width, int_bit, frac_bit):
+    '''
+    given number, width, integer bit and fractinal bit,
+    returns the fixed point representation.
+    example: decimal_to_fixed_p (11.125,8,5,3) returns 01011001 = 2^3+2^1+2^0+2^(-3)
+    precondition: There is no overflow 
+    (integer part of the number should be representable with int_bit number of bits )
+    '''
+    # seperate to int and fractional parts
+    intg , frac = str(num).split(".") 
+    int_b = np.binary_repr(int(intg), width = int_bit)
+    frac = "0."+frac
+    # multiply fractional part with 2**frac_bit to turn into integer
+    # frac= int(float(frac) * float(2**frac_bit))
+    frac = float(frac) * float(2**frac_bit)
+    _, f = str(frac).split(".") 
+    # raises Exception when the number can't be represented as fixed numbers
+    if f != "0":
+        raise Exception("number can't be represented as fixedpoint numbers")
+    frac = int(frac)
+    frac_b = np.binary_repr(frac, width = frac_bit)
+    r = int_b + frac_b 
+    return r
+
+
+def fixed_p_to_decimal (fp, width, int_bit, frac_bit):
+    '''
+    given fixedpoint representation, width, integer bit and fractinal bit,
+    returns the number.
+    example: fixed_p_to_decimal ('01011001',8,5,3) returns 11.125
+    '''
+    int_b = fp[:int_bit]
+    frac_b = fp[int_bit:width]
+    int_num = int(int_b, 2) 
+    frac = int(frac_b, 2) 
+    frac_num = float(frac / 2**(frac_bit))
+    num = float(int_num +frac_num)
+    return num
+
 
 # go through the json data and create a file for each key,
 # flattening the data and then converting it to bitstrings
+
 def convert2dat(output_dir, data, extension):
     output_dir = Path(output_dir)
     shape = {}
     for k, item in data.items():
         path = output_dir / f"{k}.{extension}"
         path.touch()
         arr = np.array(item["data"])
-        shape[k] = {"shape": list(arr.shape), "bitwidth": item["bitwidth"]}
-        with path.open('w') as f:
-            for v in arr.flatten():
-                f.write(bitstring(v, item["bitwidth"]) + "\n")
+        # type informations are given as string
+        typ = item["type"]
+        info = typ[typ.find("(")+1:typ.find(")")]
```

Comment:
This code is really brittle. Given `TFixed(...)`, it ignores the name of the type and just looks at the number of items inside the `(...)`. This means that you can write: `FooBar(1,2)`, `OhNoThisIsNotAType(3, 4, 5)` etc. and it'll do the same things.

A better solution to this is writing a function `parse_type` that takes Dahlia's surface types (`bit<32>`, `ufix<15, 8>`) and and returns a dictionary with relevant type information. For example:

1. `bit<32>`: `{ type_name: "bit", width: 32, unsigned: false }`
2.`ubit<16>`: `{ type_name: "bit", width: 16, unsigned: true }`
3.`fix<16, 8>`: `{ type_name: "fix", full_width: 16, int_width: 8, unsigned: false }`

Does that idea make sense @yoonachang? The method for parsing this should be robust—when given an invalid Dahlia type, it should show a reasonable error message.

Reply:
Yeah it makes sense, I'll make changes regarding the feedback

Reply:
@rachitnigam I have made updates according to the feedback, can you review it whenever you have time?